examples/gcoap/client.c

```diff
@@ -300,43 +337,54 @@ int gcoap_cli_cmd(int argc, char **argv)
         }
 
         if (_proxied) {
-            sock_udp_ep_t tmp_remote;
-            if (sock_udp_name2ep(&tmp_remote, argv[apos]) != 0) {
-                return _print_usage(argv);
-            }
-
-            if (tmp_remote.port == 0) {
-                if (IS_USED(MODULE_GCOAP_DTLS)) {
-                    tmp_remote.port = CONFIG_GCOAPS_PORT;
-                }
-                else {
-                    tmp_remote.port = CONFIG_GCOAP_PORT;
-                }
-            }
-
 #ifdef SOCK_HAS_IPV6
             char addrstr[IPV6_ADDR_MAX_STR_LEN];
 #else
             char addrstr[IPV4_ADDR_MAX_STR_LEN];
 #endif
-            inet_ntop(tmp_remote.family, &tmp_remote.addr, addrstr, sizeof(addrstr));
+            inet_ntop(remote.family, &remote.addr, addrstr, sizeof(addrstr));
 
-            if (tmp_remote.family == AF_INET6) {
+            if (remote.family == AF_INET6) {
                 uri_len = snprintf(proxy_uri, sizeof(proxy_uri), "coap://[%s]:%d%s",
-                                   addrstr, tmp_remote.port, uri);
+                                   addrstr, remote.port, uri);
             }
             else {
                 uri_len = snprintf(proxy_uri, sizeof(proxy_uri), "coap://%s:%d%s",
-                                   addrstr, tmp_remote.port, uri);
+                                   addrstr, remote.port, uri);
             }
 
             uri = proxy_uri;
-
-            gcoap_req_init(&pdu, &buf[0], CONFIG_GCOAP_PDU_BUF_SIZE, code_pos, NULL);
         }
-        else {
-            gcoap_req_init(&pdu, &buf[0], CONFIG_GCOAP_PDU_BUF_SIZE, code_pos, uri);
+
+        gcoap_req_init(&pdu, buf, CONFIG_GCOAP_PDU_BUF_SIZE, code_pos,
+                       (_proxied || observe) ? NULL : uri);
+
+        if (observe) {
+            uint8_t *token = coap_get_token(&pdu);
+            if (obs_value == COAP_OBS_REGISTER) {
+                obs_req_tkl = coap_get_token_len(&pdu);
+                /* backup the token of the initial observe registration */
+                memcpy(obs_req_token, token, obs_req_tkl);
+                observing = true;
+            } else {
+                /* use the token of the registration for deregistration */
+                memcpy(token, obs_req_token, obs_req_tkl);
+                if (gcoap_obs_req_forget(&remote, obs_req_token, obs_req_tkl)) {
+                    printf("could not remove observe request\n");
+                    return 1;
+                }
+                observing = false;
+                /* deregistration does not require further actions */
+                return 0;
+            }
+            coap_opt_add_uint(&pdu, COAP_OPT_OBSERVE, obs_value);
+            if (!_proxied) {
+                /* add uri path option separately
+                 * (options must be added in order) */
+                coap_opt_add_string(&pdu, COAP_OPT_URI_PATH, uri, '/');
+            }
         }
```

Comment:
Move the lower code block into the first `if` as all flow of execution returns in the `else` block anyways.
```suggestion
        if (observe) {
            uint8_t *token = coap_get_token(&pdu);
            if (obs_value == COAP_OBS_REGISTER) {
                obs_req_tkl = coap_get_token_len(&pdu);
                /* backup the token of the initial observe registration */
                memcpy(obs_req_token, token, obs_req_tkl);
                observing = true;
                coap_opt_add_uint(&pdu, COAP_OPT_OBSERVE, obs_value);
                if (!_proxied) {
                    /* add uri path option separately
                     * (options must be added in order) */
                    coap_opt_add_string(&pdu, COAP_OPT_URI_PATH, uri, '/');
                }
            } else {
                /* use the token of the registration for deregistration */
                memcpy(token, obs_req_token, obs_req_tkl);
                if (gcoap_obs_req_forget(&remote, obs_req_token, obs_req_tkl)) {
                    printf("could not remove observe request\n");
                    return 1;
                }
                observing = false;
                /* deregistration does not require further actions */
                return 0;
            }
        }
```

Reply:
makes sense. Done